tests/test_class_src.py

```diff
@@ -0,0 +1,266 @@
+import logging
+import os
+from heapq import heappush, heappushpop
+from itertools import product, repeat
+
+import numpy as np
+import pytest
+
+from aspire.basis import FFBBasis2D
+from aspire.classification import (
+    BandedSNRImageQualityFunction,
+    BFRAverager2D,
+    BumpWeightedContrastImageQualityFunction,
+    ContrastClassSelector,
+    ContrastImageQualityFunction,
+    ContrastWithRepulsionClassSelector,
+    DistanceClassSelector,
+    GlobalClassSelector,
+    GlobalWithRepulsionClassSelector,
+    RampWeightedContrastImageQualityFunction,
+    RandomClassSelector,
+    RIRClass2D,
+    TopClassSelector,
+)
+from aspire.classification.class_selection import _HeapItem
+from aspire.denoising import DebugClassAvgSource, DefaultClassAvgSource
+from aspire.image import Image
+from aspire.source import Simulation
+from aspire.utils import Rotation
+from aspire.volume import Volume
+
+logger = logging.getLogger(__name__)
+
+
+DATA_DIR = os.path.join(os.path.dirname(__file__), "saved_test_data")
+
+
+IMG_SIZES = [
+    32,
+    pytest.param(31, marks=pytest.mark.expensive),
+]
+DTYPES = [
+    np.float64,
+    pytest.param(np.float32, marks=pytest.mark.expensive),
+]
+CLS_SRCS = [DebugClassAvgSource, DefaultClassAvgSource]
+# For very small problems, it usually isn't worth running in parallel.
+NUM_PROCS = 1
+
+
+def sim_fixture_id(params):
+    res = params[0]
+    dtype = params[1]
+    return f"res={res}, dtype={dtype}"
+
+
+@pytest.fixture(params=DTYPES, ids=lambda x: f"dtype={x}")
+def dtype(request):
+    return request.param
+
+
+@pytest.fixture(params=IMG_SIZES, ids=lambda x: f"img_size={x}")
+def img_size(request):
+    return request.param
+
+
+@pytest.fixture
+def class_sim_fixture(dtype, img_size):
+    """
+    Construct a Simulation with explicit viewing angles forming
+    synthetic classes.
+    """
+
+    # Configuration
+    n_inplane_rots = 40
+
+    # Platonic solids can generate our views.
+    # Start with a cube, 8 vertices (use +-1 wlog),
+    # each represents an viewing axis.
+    cube_vertices = list(product(*repeat((-1, 1), 3)))
+    inplane_rots = np.linspace(0, 2 * np.pi, n_inplane_rots, endpoint=False)
+    # We want the first rotation to have angle 2pi instead of 0,
+    # so the norm isn't degenerate (0) later.
+    inplane_rots[0] = 2 * np.pi
+    logger.debug(f"inplane_rots: {inplane_rots}")
+
+    # Total rotations will be number of axis  * number of angles
+    # ie. vertices * n_inplane_rots
+    n = len(cube_vertices) * n_inplane_rots
+    logger.debug(f"Constructing {n} rotations.")
+
+    # Generate Rotations
+    # Normalize the rotation axes to 1
+    rotvecs = cube_vertices / np.linalg.norm(cube_vertices, axis=0)
+    logger.debug(f"rotvecs: {rotvecs}")
+    # renormalize by broadcasting with angle amounts in inplane_rots
+    rotvecs = (rotvecs[np.newaxis].T * inplane_rots).T.reshape(n, 3)
+    # Construct rotation object
+    true_rots = Rotation.from_rotvec(rotvecs, dtype=dtype)
```

Comment:
😃

Reply:
I'm assuming the smily is from seeing someone using your code.

That is a good feeling that you should cherish.

Reply:
Yep, thanks!